### PR TITLE
8316235: Optimization for DateTimeFormatter::format

### PR DESCRIPTION
**Pull request by wenshao**
In many scenarios, DateTimeFormatter::format is a slower operation.  

For example, the following business scenarios
1. The json library gson/jackson/[fastjson2](https://github.com/alibaba/fastjson2) formats Instant/LocalDate/LocalTime/LocalDateTime/ZonedDateTim into strings.
2. In data integration scenarios, for projects like  [datax](https://github.com/alibaba/datax)/[canal](https://github.com/alibaba/canal), if the input type is Date/Instant and the output type is String, formatting is required.

This PR provides format performance optimization for commonly used date patterns.
```
ISO_INSTANT
ISO_LOCAL_TIME
ISO_LOCAL_DATE
ISO_LOCAL_DATETIME
HH:mm:ss
HH:mm:ss.SSS
yyyy-MM-dd
yyyy-MM-dd HH:mm:ss
yyyy-MM-dd'T'HH:mm:ss
yyyy-MM-dd HH:mm:ss.SSS
yyyy-MM-dd'T'HH:mm:ss.SSS
```

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8316235](https://bugs.openjdk.org/browse/JDK-8316235): Optimization for DateTimeFormatter::format (**Enhancement** - P4)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15722/head:pull/15722` \
`$ git checkout pull/15722`

Update a local copy of the PR: \
`$ git checkout pull/15722` \
`$ git pull https://git.openjdk.org/jdk.git pull/15722/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15722`

View PR using the GUI difftool: \
`$ git pr show -t 15722`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15722.diff">https://git.openjdk.org/jdk/pull/15722.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15722#issuecomment-1718650068)

### CODE REVIEW

**Comment by wenshao** [comment]
Performance comparison data is as follows:

## 1. Script
```
bash configure
make images
sh make/devkit/createJMHBundle.sh
bash configure --with-jmh=build/jmh/jars
make test TEST="micro:java.time.format.DateTimeFormatterBench.*"
```

## 2. MacBookPro benchmark 
```diff
-Benchmark                                                   (pattern)   Mode  Cnt   Score   Error   Units (baseline)
-DateTimeFormatterBench.formatInstants                        HH:mm:ss  thrpt   15  14.888 ? 0.109  ops/ms
-DateTimeFormatterBench.formatInstants                    HH:mm:ss.SSS  thrpt   15  10.132 ? 0.046  ops/ms
-DateTimeFormatterBench.formatInstants           yyyy-MM-dd'T'HH:mm:ss  thrpt   15   8.993 ? 0.039  ops/ms
-DateTimeFormatterBench.formatInstants       yyyy-MM-dd'T'HH:mm:ss.SSS  thrpt   15   7.400 ? 0.035  ops/ms
-DateTimeFormatterBench.formatZonedDateTime                   HH:mm:ss  thrpt   15  21.460 ? 0.056  ops/ms
-DateTimeFormatterBench.formatZonedDateTime               HH:mm:ss.SSS  thrpt   15  14.439 ? 0.264  ops/ms
-DateTimeFormatterBench.formatZonedDateTime      yyyy-MM-dd'T'HH:mm:ss  thrpt   15  12.742 ? 0.055  ops/ms
-DateTimeFormatterBench.formatZonedDateTime  yyyy-MM-dd'T'HH:mm:ss.SSS  thrpt   15   9.059 ? 0.124  ops/ms

+Benchmark                                                   (pattern)   Mode  Cnt   Score   Error   Units (optimized)
+DateTimeFormatterBench.formatInstants                        HH:mm:ss  thrpt   15  28.082 ? 0.284  ops/ms (+88.62%)
+DateTimeFormatterBench.formatInstants                    HH:mm:ss.SSS  thrpt   15  25.483 ? 0.109  ops/ms (+151.51%)
+DateTimeFormatterBench.formatInstants           yyyy-MM-dd'T'HH:mm:ss  thrpt   15  19.950 ? 0.444  ops/ms (+121.83%)
+DateTimeFormatterBench.formatInstants       yyyy-MM-dd'T'HH:mm:ss.SSS  thrpt   15  18.391 ? 0.327  ops/ms (+148.52%)
+DateTimeFormatterBench.formatZonedDateTime                   HH:mm:ss  thrpt   15  57.870 ? 0.641  ops/ms (+169.66%)
+DateTimeFormatterBench.formatZonedDateTime               HH:mm:ss.SSS  thrpt   15  43.783 ? 0.300  ops/ms (+203.22%)
+DateTimeFormatterBench.formatZonedDateTime      yyyy-MM-dd'T'HH:mm:ss  thrpt   15  36.220 ? 0.194  ops/ms (+184.25%)
+DateTimeFormatterBench.formatZonedDateTime  yyyy-MM-dd'T'HH:mm:ss.SSS  thrpt   15  32.512 ? 0.583  ops/ms (+258.89%)
```

File: src/java.base/share/classes/java/time/format/DateTimeFormatterBuilder.java [hunk 1]
```diff
@@ -2488,15 +2561,130 @@ interface DateTimePrinterParser {
...
+                date = ((LocalDateTime) temporal).toLocalDate();
+            } else if (temporal instanceof LocalDate) {
+                date = (LocalDate) temporal;
+            } else if (temporal instanceof ZonedDateTime) {
+                date = ((ZonedDateTime) temporal).toLocalDate();
+            } else if (temporal instanceof OffsetDateTime) {
+                date = ((OffsetDateTime) temporal).toLocalDate();
+            } else {
+                return super.format(context, buf);
+            }
```
Comment (ExE-Boss) [inline src/java.base/share/classes/java/time/format/DateTimeFormatterBuilder.java:2594]: This can use `instanceof <pattern>`:
```suggestion
            if (temporal instanceof LocalDateTime ldt) {
                date = ldt.toLocalDate();
            } else if (temporal instanceof LocalDate ld) {
                date = ld;
            } else if (temporal instanceof ZonedDateTime zdt) {
                date = zdt.toLocalDate();
            } else if (temporal instanceof OffsetDateTime odt) {
                date = odt.toLocalDate();
            } else {
                return super.format(context, buf);
            }
```

Or even a pattern switch:
```suggestion
            switch (temporal) {
                case LocalDateTime ldt -> date = ldt.toLocalDate();
                case LocalDate ld -> date = ld;
                case ZonedDateTime zdt -> date = zdt.toLocalDate();
                case OffsetDateTime odt -> date = odt.toLocalDate();
                default -> {
                    return super.format(context, buf);
                }
            }
```

File: src/java.base/share/classes/java/time/format/DateTimeFormatterBuilder.java [hunk 2]
```diff
@@ -2488,15 +2561,130 @@ interface DateTimePrinterParser {
...
+                time = ((LocalDateTime) temporal).toLocalTime();
+            } else if (temporal instanceof ZonedDateTime) {
+                time = ((ZonedDateTime) temporal).toLocalTime();
+            } else if (temporal instanceof OffsetDateTime) {
+                time = ((OffsetDateTime) temporal).toLocalTime();
+            } else if (temporal instanceof OffsetTime) {
+                time = ((OffsetTime) temporal).toLocalTime();
+            } else {
+                return super.format(context, buf);
+            }
```
Comment (ExE-Boss) [inline src/java.base/share/classes/java/time/format/DateTimeFormatterBuilder.java:2661]: This can use `instanceof <pattern>`:
```suggestion
            if (temporal instanceof LocalTime lt) {
                time = lt;
            } else if (temporal instanceof LocalDateTime ldt) {
                time = ldt.toLocalTime();
            } else if (temporal instanceof ZonedDateTime zdt) {
                time = zdt.toLocalTime();
            } else if (temporal instanceof OffsetDateTime odt) {
                time = odt.toLocalTime();
            } else if (temporal instanceof OffsetTime ot) {
                time = ot.toLocalTime();
            } else {
                return super.format(context, buf);
            }
```

Or even a pattern switch:
```suggestion
            switch (temporal) {
                case LocalTime lt -> time = lt;
                case LocalDateTime ldt -> time = ldt.toLocalTime();
                case ZonedDateTime zdt -> time = zdt.toLocalTime();
                case OffsetDateTime odt -> time = odt.toLocalTime();
                case OffsetTime ot -> time = ot.toLocalTime();
                default -> {
                    return super.format(context, buf);
                }
            }
```

File: src/java.base/share/classes/java/time/format/DateTimeFormatterBuilder.java [hunk 3]
```diff
@@ -2488,15 +2561,130 @@ interface DateTimePrinterParser {
...
+            return true;
+        }
+    }
+
     //-----------------------------------------------------------------------
     /**
      * Composite printer and parser.
      */
-    static final class CompositePrinterParser implements DateTimePrinterParser {
+    static class CompositePrinterParser implements DateTimePrinterParser {
```
Comment (ExE-Boss) [inline src/java.base/share/classes/java/time/format/DateTimeFormatterBuilder.java:2683]: This class can be `sealed`:
```suggestion
    static sealed class CompositePrinterParser implements DateTimePrinterParser {
```

Comment (wenshao) [reply src/java.base/share/classes/java/time/format/DateTimeFormatterBuilder.java]: Changes have been made based on your suggestions

File: src/java.base/share/classes/java/time/format/DateTimeFormatterBuilder.java (hunk 2 above)
Comment (wenshao) [reply src/java.base/share/classes/java/time/format/DateTimeFormatterBuilder.java]: It's a good idea to use the new switch syntax

**Pull request closed without merging**